src/ogdf/basic/graph_generators/operations.cpp

```diff
@@ -246,4 +250,117 @@ void rootedProduct(const Graph& G1, const Graph& G2, Graph& product, NodeMap& no
 	});
 }
 
+void complement(Graph& G, bool directional, bool allow_self_loops) {
+	NodeSet<true> n1neighbors(G);
+	EdgeSet<true> newEdges(G);
+
+	for (node n1 : G.nodes) {
+		// deleting edges
+		safeForEach(n1->adjEntries, [&](adjEntry adj) {
+			node n2 = adj->twinNode();
+
+			if (directional && !adj->isSource()) {
+				return;
+			}
+			if (!directional && n1->index() > n2->index()) {
+				return;
+			}
+			if (newEdges.isMember(adj->theEdge())) {
+				return;
+			}
+			n1neighbors.insert(n2);
+			G.delEdge(adj->theEdge());
+		});
+
+		// adding edges
+		for (node n2 : G.nodes) {
+			if (!directional && n1->index() > n2->index()) {
+				continue;
+			}
+			if (!allow_self_loops && n1->index() == n2->index()) {
+				continue;
+			}
+			if (n1neighbors.isMember(n2)) {
+				continue;
+			}
+
+			edge newEdge = G.newEdge(n1, n2);
+			newEdges.insert(newEdge);
+		}
+		n1neighbors.clear();
+	}
+}
+
+void intersection(Graph& G1, const Graph& G2, const NodeArray<node>& nodeMap) {
+	OGDF_ASSERT(nodeMap.valid());
+
+	safeForEach(G1.nodes, [&](node n1) {
+		node n2 = nodeMap[n1];
+		if (n2 == nullptr) {
+			G1.delNode(n1);
+		}
+	});
+
+	if (isParallelFree(G1)) { // no multi-edges
+		for (node n1a : G1.nodes) {
+			node n2a = nodeMap[n1a];
+			List<edge> edgelist;
+			n1a->adjEdges(edgelist);
+
+			for (edge e1 : edgelist) {
+				node n1b = e1->opposite(n1a);
+				node n2b = nodeMap[n1b];
+				edge e2 = G2.searchEdge(n2a, n2b);
+
+				if (e2 == nullptr) {
+					G1.delEdge(e1);
+				}
+			}
+		}
+	} else {
+		for (node n1a : G1.nodes) {
+			node n2a = nodeMap[n1a];
+			List<edge> edgelist;
+			n1a->adjEdges(edgelist);
+
+			EdgeArray<SListPure<edge>> edgeArr(G1);
+			getParallelFreeUndirected(G1, edgeArr);
+
+
+			for (edge e1 : edgelist) {
+				node n1b = e1->opposite(n1a);
+				node n2b = nodeMap[n1b];
+				edge e2 = G2.searchEdge(n2a, n2b);
+
+				EdgeArray<SListPure<edge>> edgeArr2(G1);
+				getParallelFreeUndirected(G1, edgeArr2);
```

Comment:
This again makes the runtime super-quadratic and in practice is even worse as it goes through all edges and not only those of a single node, while the intersection can surely be found in quadratic time total. Why don't you simply use `AdjacencyOracle`, i.e. build it once for G2 and then simply delete all edges in G1 for which the oracle returns false? This should also entirely safe you the case distinction with parallel edges.

Reply:
Looks better now, but the two getParallelFreeUndirect calls still break the runtime bound without actually having an effect. Why do you need to handle graphs with parallel edges differently anyways?